core/test/base/reflection.cpp

```diff
@@ -0,0 +1,141 @@
+
+#include <ginkgo/core/base/reflection.hpp>
+
+
+#include <vector>
+
+
+#include <gtest/gtest.h>
+
+
+namespace {
+
+
+struct DummyOperator : public gko::EnableLinOp<DummyOperator> {
+    DummyOperator(std::shared_ptr<const gko::Executor> exec,
+                  gko::dim<2> size = {})
+        : gko::EnableLinOp<DummyOperator>(exec, size)
+    {}
+
+    void apply_impl(const LinOp *b, LinOp *x) const override {}
+
+    void apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                    LinOp *x) const override
+    {}
+};
+
+
+struct TransposableDummyOperator
+    : public gko::EnableLinOp<TransposableDummyOperator>,
+      public gko::Transposable {
+    TransposableDummyOperator(std::shared_ptr<const gko::Executor> exec,
+                              gko::dim<2> size = {})
+        : gko::EnableLinOp<TransposableDummyOperator>(exec, size)
+    {}
+
+    void apply_impl(const LinOp *b, LinOp *x) const override {}
+
+    void apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                    LinOp *x) const override
+    {}
+
+    std::unique_ptr<LinOp> transpose() const override
+    {
+        auto result = std::unique_ptr<TransposableDummyOperator>(
+            new TransposableDummyOperator(this->get_executor(),
+                                          gko::transpose(this->get_size())));
+        return std::move(result);
+    }
+
+    std::unique_ptr<LinOp> conj_transpose() const override
+    {
+        auto result = this->transpose();
+        return std::move(result);
+    }
+};
+
+
+class Reflection : public ::testing::Test {
+protected:
+    Reflection()
+        : exec{gko::ReferenceExecutor::create()},
+          U{std::make_shared<DummyOperator>(exec, gko::dim<2>{2, 1})},
+          V{std::make_shared<DummyOperator>(exec, gko::dim<2>{1, 2})},
+          transU{std::make_shared<TransposableDummyOperator>(
+              exec, gko::dim<2>{3, 1})},
+          coef{std::make_shared<DummyOperator>(exec, gko::dim<2>{1, 1})}
+    {}
+
+    std::shared_ptr<const gko::Executor> exec;
+    std::shared_ptr<gko::LinOp> U;
+    std::shared_ptr<gko::LinOp> V;
+    std::shared_ptr<gko::LinOp> transU;
+    std::shared_ptr<gko::LinOp> coef;
+};
+
+
+TEST_F(Reflection, CanBeEmpty)
+{
+    auto cmp = gko::Reflection<>::create(exec);
+
+    ASSERT_EQ(cmp->get_size(), gko::dim<2>(0, 0));
+}
+
+
+TEST_F(Reflection, CanCreateFromTwoOperators)
+{
+    auto cmp = gko::Reflection<>::create(coef, U, V);
+
+    ASSERT_EQ(cmp->get_size(), gko::dim<2>(2, 2));
+    ASSERT_EQ(cmp->get_u_operator(), U);
+    ASSERT_EQ(cmp->get_v_operator(), V);
+}
+
+
+TEST_F(Reflection, CanNotCreateFromOneNonTransableOperators)
+{
+    ASSERT_THROW(gko::Reflection<>::create(coef, U), gko::NotSupported);
+}
+
+
+TEST_F(Reflection, CanCreateFromOneTransableOperators)
+{
+    auto cmp = gko::Reflection<>::create(coef, transU);
+    ASSERT_EQ(cmp->get_size(), gko::dim<2>(3, 3));
```

Comment:
Newline before the `ASSERT` to separate each A in AAA (Arrange, Act, Assert) with a newline, according to our [guidelines](https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines#whitespace).